protocols/protocol/cloudstate/entity.proto

```diff
@@ -69,41 +109,70 @@ message SideEffect {
     // The payload of the command.
     google.protobuf.Any payload = 3;
 
-    // Whether this side effect should be performed synchronously, ie, before the reply is eventually
-    // sent, or not.
+    // Whether this side effect should be performed synchronously, ie, before the reply is eventually sent, or not.
     bool synchronous = 4;
 }
 
-// A command. For each command received, a reply must be sent with a matching command id.
+//Command represents TODO
+//
+//---
+//
+//`Specification`:
+//
+//  - An `Command` is considered to be valid iff:
+//    - `entity_id` TODO
+//    - `id` TODO
+//    - `name` TODO
+//    - `payload` TODO
+//    - `streamed` TODO
 message Command {
 
     // The ID of the entity.
     string entity_id = 1;
 
     // A command id.
+    // @exclude FIXME rename to command_id
     int64 id = 2;
 
-    // Command name
+    // Command name.
     string name = 3;
 
     // The command payload.
     google.protobuf.Any payload = 4;
 
-    // Whether the command is streamed or not
+    // Whether the command is streamed or not.
     bool streamed = 5;
 }
 
+//StreamCancelled TODO
+//
+//---
+//
+//`Specification`:
+//
+//  - A `StreamCancelled` is considered to be valid iff:
+//    - `entity_id` TODO
+//    - `id` TODO
 message StreamCancelled {
 
-    // The ID of the entity
+    // The ID of the entity.
     string entity_id = 1;
 
-    // The command id
+    // The command id.
+    // @exclude FIXME Rename to command_id
     int64 id = 2;
 }
 
-// A failure reply. If this is returned, it will be translated into a gRPC unknown
-// error with the corresponding description if supplied.
+//Failure represents a Command processing failure reply.
+//If this is returned, it will be translated into a gRPC unknown error with the corresponding description if supplied.
+//
+//---
+//
+//`Specification`:
+//
+//  - An `Failure` is considered to be valid iff:
+//    - `command_id` TODO
+//    - `description` TODO
 message Failure {
 
     // The id of the command being replied to. Must match the input command.
```

Comment:
We might mention that the `command_id` is optional and has be to be set (leaving it with its default value of 0) if a `Failure` is sent as a reply not according to an input command.

Reply:
Line 166 states that "Failure represents a Command processing failure reply." In this sense, this Failure Message  can't be used for anything not command related. Is that correct?

Reply:
@marcellanz Good point, worth clarifying.

Reply:
`Failure` is used in two places.

The first place is as part of `ClientAction`. When sent as a client action, it indicates an error that should be sent to the client - this is an intentional, explicit error message that the user function wants to communicate with the client. In that case, command id is actually irrelevant, since it's already included in a reply message that contains the command id. Typically, this will be used for errors made by the client, for example, commands that failed validation, and could be thought of as an HTTP 4xx error.

The second place that `Failure` is used is by itself, as a top level message in the oneof for a stream out for the event sourcing, crdt, etc protocols. In this case, it's used for server errors, logically equivalent to HTTP 5xx errors, whose message should not be communicated back to the client (since the client hasn't done anything wrong and can't do anything to rectify, so the error message will be meaningless to the client, plus leaking error messages from arbitrary exceptions can be a security issue), rather a server error status code with a generic error message is sent, but the error will be logged by the proxy for diagnostic purposes. These may or may not have a command id. For example, if the users code throws an unexpected exception, or does something illegal according to the support library, this will be used, and if that occurred during the processing of a command, it will have a command id. However, if the error occurred outside of command handling, it won't have a command id. An example of where we currently send it without a command id is if the proxy sends two init messages (that's not allowed) the user function support library responds with this failure message and no command id. When this is sent, the user function and the proxy are both expected to immediately close the stream after sending/receiving it, since it's not safe to make any assumptions about the state of the stream, the state of the other end of the stream, and any local state related to the stream, when such an error occurs.

Does it make sense to use the same message for both purposes? Maybe, maybe not.

Reply:
great explanations, if this is well documented I don't see why segregate into more messages

Reply:
@jroper thanks for the explanation.

Reply:
>When this is sent, the user function and the proxy are both expected to immediately close the stream after sending/receiving it, since it's not safe to make any assumptions about the state of the stream, the state of the other end of the stream, and any local state related to the stream, when such an error occurs.

This would be similar to a `ClientAction.Failure`? After such an intentional error, we can't make any assumptions about the local state, and probably too would close the stream and let the proxy re-establish the local state if he likes to.